app/controllers/api/groups_controller.rb

```diff
@@ -80,5 +80,54 @@ def include_students(fields)
       end
     end
 
+    # Update the group's marks for the given assignment.
+    def update_marks
+      assignment = Assignment.find_by_id(params[:assignment_id])
+      if assignment.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No assignment exists with that id' }, status: 404
+        return
+      end
+
+      group = Group.find_by_id(params[:id])
+      if group.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No group exists with that id' }, status: 404
+        return
+      elsif group.submissions.first.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No submissions exist for that group' }, status: 404
+        return
+      end
+
+      all_criteria = assignment.rubric_criteria
+      all_criteria.push(*assignment.flexible_criteria)
+      if all_criteria.empty?
+        render 'shared/http_status', locals: { code: '404', message:
+          'No criteria exist for that assignment' }, status: 404
+        return
+      end
+
+      result = group.submissions.first.get_latest_result
+      params.keys.each do |param_key|
```

Comment:
This loop structure is unnecessary.

Set up a query like `RubricCriterion.where(assignment: assignment, rubric_criterion_name: params[:name])` to do the search. (Need another one for FlexibleCriterion too.)